src/Chapter3.hs

```diff
@@ -1006,6 +1131,33 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+-- knight [ health, attack, defense]
+-- - attack
+-- - drink health potion 
+-- - cast spell to incraese defense 
+
+-- monster [ health, attack ] 
+-- - attack 
+-- - run 
+
+data T9_KnightAction = KnightAttack | KnightDrink | KnightCast deriving (Show)
+data T9_MonsterAction = MonsterAttack | MonsterRun deriving (Show)
+
+data T9_Knight = Knight {
+  knightHealth :: Int,
+  knightAttack :: Int,
+  knightDefense :: Int
+} deriving (Show)
+
+data T9_Monster = Monster {
+  monsteHealth :: Int,
+  monsterAttack :: Int
+}
+class Fighter a where
+  attack :: Int
+  health :: Int
```

Comment:
That's a good start! I can provide some hints for this task to help you move forward 🙂 

For this task you need to create two data types: `Monster` and `Knight` (as you already correctly did). Each type will have the properties described in the task description.

In addition, I would recommend keeping separate types for `MonsterAction` and `KnightAction` to represent possible actions for each type correspondingly.

Since each data type (`Monster` and `Knight`) can be fighters, I can suggest creating a typeclass of the following shape:

```haskell
class Fighter a where
    ... you need to think of possible methods here ...
```

So you can them implement a recursive function using this typeclass:

```haskell

data FightResult
    = FirstWins
    | SecondWins

fight :: (Fighter a, Fighter b) => a -> b -> FightResult
```

Your current typeclass methods don't take values of type `a` as arguments. Which means that they are constant for all values. To patch the typeclass definition, you need to write it like this:

```haskell
class Fighter a where
  attack :: a -> Int
  health :: a -> Int
```

See, how methods are functions from values to fields you want to get? You can try implementing a generic `Fight` function, so you will be able to see, what methods you need to add to the `Fighter` typeclass.

I hope this clarifies the task a bit more and gives some ideas on how to proceed with it :slightly_smiling_face: